src/Computation/Expression.cs

```diff
@@ -2,308 +2,500 @@
 using System.Collections.Generic;
 using System.Linq;
 using System.Numerics;
+using Microsoft.Z3;
 using Symbolica.Collection;
+using Symbolica.Computation.Exceptions;
+using Symbolica.Computation.Values;
+using Symbolica.Computation.Values.Constants;
+using Symbolica.Computation.Values.Symbolics;
 using Symbolica.Expression;
 
 namespace Symbolica.Computation
 {
     internal sealed class Expression : IExpression
     {
         private readonly ICollectionFactory _collectionFactory;
-        private readonly Lazy<BigInteger> _constant;
-        private readonly IBool[]? _constraints;
+        private readonly IValue[]? _constraints;
         private readonly IContextFactory _contextFactory;
         private readonly IValue _value;
 
         private Expression(IContextFactory contextFactory, ICollectionFactory collectionFactory,
-            IValue value, IBool[]? constraints)
+            IValue value, IValue[]? constraints)
         {
             _contextFactory = contextFactory;
             _collectionFactory = collectionFactory;
             _value = value;
             _constraints = constraints;
-            _constant = new Lazy<BigInteger>(AsConstant);
         }
 
         public Bits Size => _value.Size;
-        public BigInteger Constant => _constant.Value;
+
+        public BigInteger Constant => _value is IConstantValue c
+            ? c.AsUnsigned()
+            : AsConstant(_contextFactory, _value);
 
         public IExpression GetValue(ISpace space)
         {
-            return Create(c => c.GetValue((IPersistentSpace) space, _constraints ?? Array.Empty<IBool>()));
+            return Unary(
+                v => GetValue((IPersistentSpace) space, v, _constraints ?? Array.Empty<IValue>()),
+                v => v);
         }
 
         public IProposition GetProposition(ISpace space)
         {
-            return _value.AsBool().GetProposition((IPersistentSpace) space, _constraints ?? Array.Empty<IBool>());
+            return _value is IConstantValue c
```

Comment:
This type switching in here, to decide if something is constant or symbolic, made me wonder if we should have two implementations of `IExpression`, a `ConstantExpression` and a `SymbolicExpression`. You've probably already thought about this though, so just curious why you chose this design?

Reply:
I have switched over to this design now. I had something similar to this way back but other issues caused a problem with it. These are no longer valid though so I think it is better.